superset-frontend/src/explore/components/AdhocMetricEditPopover.jsx

```diff
@@ -100,10 +100,13 @@ export default class AdhocMetricEditPopover extends React.Component {
   }
 
   onSave() {
-    // unset isNew here in case save button was clicked when no changes were made
+    if (this.props.title.hasCustomLabel) {
+      this.state.adhocMetric.label = this.props.title.label;
+      this.state.adhocMetric.hasCustomLabel = true;
+    }
```

Comment:
To avoid mutating state directly, could we perhaps create a new object with `label` and `hasCustomLabel` here and then spread that in the `onChange` object?

Reply:
Sorry, I am a bit new to these jsx conventions. Could you specify the rule which suggests avoiding mutating states?

Reply:
Is this the way you want?
```
  onSave() {
    const label = this.props.title.hasCustomLabel ?
      this.props.title.label : this.state.adhocMetric.label;
    const hasCustomLabel = this.props.title.hasCustomLabel;
    this.props.onChange({
      ...this.state.adhocMetric,
      label,
      hasCustomLabel
    });
  }
```